pkg/grpc/interceptor/auth/server.go

```diff
@@ -26,41 +39,99 @@ type authData struct {
 	team      string
 }
 
-func extractAuthFromContext(ctx context.Context) (*authData, error) {
+func NewServerInterceptor(apiKeyStore database.ApiKeyStore, tokenValidator TokenValidator, teamsClient TeamsClient) *ServerInterceptor {
+	prometheus.MustRegister(promReq)
+	return &ServerInterceptor{
+		APIKeyStore:    apiKeyStore,
+		TokenValidator: tokenValidator,
+		TeamsClient:    teamsClient,
+	}
+}
+
+var (
+	promReq = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "hookd_auth_interceptor_requests",
+			Help: "Number of requests by type in auth interceptor",
+		},
+		[]string{"type"})
+	promErr = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "hookd_auth_interceptor_errors",
+			Help: "Number of errors in auth interceptor",
+		},
+		[]string{"type"})
+)
+
+func (s *ServerInterceptor) UnaryServerInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	_, ok := req.(*pb.DeploymentRequest)
+	if !ok {
+		return nil, status.Errorf(codes.InvalidArgument, "requests to this endpoint must be DeploymentRequest")
+	}
+
 	md, ok := metadata.FromIncomingContext(ctx)
 	if !ok {
-		return nil, status.Errorf(codes.Unauthenticated, "request is not signed with API key; metadata is missing from request headers")
+		return nil, status.Errorf(codes.Unauthenticated, "invalid metadata in request")
 	}
 
-	hmac := md["authorization"]
-	if len(hmac) == 0 {
-		return nil, status.Errorf(codes.Unauthenticated, "request is not signed with API key")
+	jwt := get("jwt", md)
+
+	if jwt != "" {
+		promReq.WithLabelValues("jwt").Inc()
+		t, err := s.TokenValidator.Validate(jwt)
+		if err != nil {
+			promErr.WithLabelValues("invalid_jwt").Inc()
+			return nil, status.Errorf(codes.Unauthenticated, "invalid JWT token")
+		}
+
+		r, ok := t.Get("repository")
+		if !ok {
+			return nil, status.Errorf(codes.Unauthenticated, "missing repository in JWT token")
+		}
+		repo := r.(string)
+
+		team := get("team", md)
+		if team == "" {
+			return nil, status.Errorf(codes.Unauthenticated, "missing team in metadata")
+		}
+
+		if s.TeamsClient.IsAuthorized(repo, team) {
+			return handler(ctx, req)
```

Comment:
Breaking out like this in the middle of a decently sized function can be really hard to spot when reading the code, and quickly results in errors when we re-visit this code.
Can we move the "check if auth if OK" logic to it's own function(s) and pass the request to the next handler *only* in the bottom of this function?